lib/saluki-components/src/destinations/datadog_metrics/request_builder.rs

```diff
@@ -243,11 +256,79 @@ where
 
         self.compression_estimator.reset();
 
+        // Clear scratch buf data.
+        let scratch_buf = std::mem::take(&mut self.scratch_buf);
+        let scratch_buf_lens = std::mem::take(&mut self.scratch_buf_lens);
+
         let new_compressor = create_compressor(&self.buffer_pool).await;
         let mut compressor = std::mem::replace(&mut self.compressor, new_compressor);
-        compressor.shutdown().await.context(Io)?;
+        if let Err(e) = compressor.shutdown().await.context(Io) {
+            return vec![Err(e)];
+        }
+
         let buffer = compressor.into_inner();
 
+        let compressed_len = buffer.len();
+        let compressed_limit = self.endpoint.compressed_size_limit();
+        if compressed_len > compressed_limit {
+            return self.split_request_efficient(scratch_buf, scratch_buf_lens).await;
+        }
+
+        debug!(endpoint = ?self.endpoint, uncompressed_len, compressed_len, "Flushing request.");
+
+        vec![self.create_request(buffer).map(|req| (metrics_written, req))]
+    }
+
+    async fn split_request_efficient(
+        &self, scratch_buf: Vec<u8>, scratch_buf_lens: Vec<usize>,
+    ) -> Vec<Result<(usize, Request<ReplayBody<ChunkedBuffer<O>>>), RequestBuilderError>> {
+        let mut requests = Vec::new();
+
+        if scratch_buf_lens.is_empty() {
+            return requests;
+        }
+
+        let pivot = scratch_buf_lens.len() / 2;
+        let first_half: usize = scratch_buf_lens.iter().take(pivot + 1).sum();
```

Comment:
I see what you're doing, but I think it could be a little simpler.

You have four possible scenarios:

- no metrics encoded (you handle this when checking `scratch_buf_lens.is_empty()`)
- single metric encoded
- even number of metrics encoded (two or more)
- odd number of metrics encode (three or more)

If you have a single metric, you're going to end up writing no bytes to the second compressor, because `first_half` will be a number that equals the total number of bytes in the scratch buffer. If you have two metrics, you're going to do the same thing, since you always take at least two items from `scratch_buf_lens` if it has two or more elements. Once we get to three metrics, it'll start biasing the larger half towards "first".

So we can solve the problem of "single metric" by just putting a check in `flush`, when we exceed our compressed size limit, to return an error directly: if a single metric can't be compressed within the limit, we have nothing to split, and thus nothing to send.

When it comes to two or more metrics, I think a small tweak in your logic would fix things up. Specifically, if we do something like this:

```rust
let lens_pivot = scratch_buf_lens.len() / 2;
let first_half_metrics_len = scratch_buf_lens.len() - lens_pivot;
let second_half_metrics_len = scratch_buf_lens.len() - first_half_metrics_len;

let scratch_buf_pivot = scratch_buf_lens.iter().take(first_half_metrics_len).sum();

let first_half_scratch_buf = scratch_buf[0..scratch_buf_pivot];
let second_half_scratch_buf = scratch_buf[scratch_buf_pivot..];
```

Instead of always adding one to the value given to `take`, you do it implicitly by capturing the rounded-down portion of `scratch_buf_lens.len() / 2`: when `scratch_buf_lens.len()` is 2, the pivot is 1, so we only take 1, and when `scratch_buf_lens.len() / 2` is 3, the pivot is also still 1, but since we do `scratch_buf_lens.len() - pivot`, we take 2 on the first half.

The rest of the variables are just then deriving from that, and in particular, generating the relevant slices for the first and second half. We also have `first_half_metrics_len` and `second_half_metrics_len` now, which we should be using when we build the requests: currently you're passing a hard-coded value of `1` which will mess up the metrics we emit.